src/stream/src/executor/mview/materialize.rs

```diff
@@ -59,13 +59,18 @@ pub struct MaterializeExecutor<S: StateStore, SD: ValueRowSerde> {
     version_column_index: Option<u32>,
 
     may_have_downstream: bool,
+
+    depended_subscription_ids: HashSet<u32>,
 }
 
 fn get_op_consistency_level(
     conflict_behavior: ConflictBehavior,
     may_have_downstream: bool,
+    depended_subscriptions: &HashSet<u32>,
 ) -> StateTableOpConsistencyLevel {
-    if !may_have_downstream && matches!(conflict_behavior, ConflictBehavior::Overwrite) {
+    if !depended_subscriptions.is_empty() {
+        StateTableOpConsistencyLevel::LogStoreEnabled
```

Comment:
Is this the only usage for maintaining the `depended_subscriptions` field?

This reminds of https://github.com/risingwavelabs/risingwave/pull/16348#discussion_r1577302916. Now we have 3 similar usages for doing optimization based on whether there's dependent jobs:

- source backfill pauses
- on-conflict check
- materialize op consistency for subscription

It would be better to find a unified approach to address all similar problems together. Passing a `related_subscriptions` field when `build_actors` and creating two mutations specifically for subscriptions appear somewhat ad-hoc to me.

cc @xxchan @chenzl25

Reply:
> Is this the only usage for maintaining the `depended_subscriptions` field?
> 
> This reminds of [#16348 (comment)](https://github.com/risingwavelabs/risingwave/pull/16348#discussion_r1577302916). Now we have 3 similar usages for doing optimization based on whether there's dependent jobs:
> 
> * source backfill pauses
> * on-conflict check
> * materialize op consistency for subscription
> 
> It would be better to find a unified approach to address all similar problems together. Passing a `related_subscriptions` field when `build_actors` and creating two mutations specifically for subscriptions appear somewhat ad-hoc to me.
> 
> cc @xxchan @chenzl25

Indeed. I think a cleaner implementation will be meta node explicitly notify changes on op consistency by the barrier instead of maintaining the downstream information locally at the actor. 

This PR is for the new subscription implementation. We hope to include it in the coming release(code frozen at the end of April), and otherwise we can only release the current subscription  implementation and will have to maintain backward compatibility between the current one and the new one.

Can we have this PR merged first if there is no correctness issue? Barrier mutation is not persisted, and we can easily refactor to implement the unified approach in the future.

Reply:
> This PR is for the new subscription implementation.

Do we have a tracking issue for this? Are the remaining tasks for new subscription already done?

Reply:
It has three remaining PR: this one, https://github.com/risingwavelabs/risingwave/pull/16487, and a coming PR to switch from current implementation to the new implementation. The last PR has been implemented and will be submitted after the first two PRs gets merged.